core/src/apps/webauthn/credential.py

```diff
@@ -0,0 +1,313 @@
+        cred.user_id = data.get(_CRED_ID_USER_ID, None)
+        cred.user_name = data.get(_CRED_ID_USER_NAME, None)
+        cred.user_display_name = data.get(_CRED_ID_USER_DISPLAY_NAME, None)
+        cred._creation_time = data.get(_CRED_ID_CREATION_TIME, 0)
+        cred.hmac_secret = data.get(_CRED_ID_HMAC_SECRET, False)
+        cred.id = cred_id
+
+        if (
+            not cred.check_required_fields()
+            or not cred.check_data_types()
+            or hashlib.sha256(cred.rp_id).digest() != rp_id_hash
+        ):
+            return None
+
+        return cred
+
+    def check_required_fields(self) -> bool:
+        return (
+            self.rp_id is not None
+            and self.user_id is not None
+            and self._creation_time is not None
+        )
+
+    def check_data_types(self) -> bool:
+        return (
+            isinstance(self.rp_id, str)
+            and isinstance(self.rp_name, (str, type(None)))
+            and isinstance(self.user_id, (bytes, bytearray))
+            and isinstance(self.user_name, (str, type(None)))
+            and isinstance(self.user_display_name, (str, type(None)))
+            and isinstance(self.hmac_secret, bool)
+            and isinstance(self._creation_time, (int, type(None)))
+            and isinstance(self.id, (bytes, bytearray))
+        )
+
+    def app_name(self) -> str:
+        return self.rp_id
+
+    def account_name(self) -> Optional[str]:
+        from ubinascii import hexlify
+
+        if self.user_name:
+            return self.user_name
+        elif self.user_display_name:
+            return self.user_display_name
+        elif self.user_id:
+            return hexlify(self.user_id).decode()
+        else:
+            return None
+
+    def private_key(self) -> bytes:
+        from apps.common import seed
+
+        path = [HARDENED | 10022, HARDENED | int.from_bytes(self.id[:4], "big")] + [
+            HARDENED | i for i in ustruct.unpack(">4L", self.id[-16:])
+        ]
+        node = seed.derive_node_without_passphrase(path, "nist256p1")
+        return node.private_key()
+
+    def hmac_secret_key(self) -> Optional[bytes]:
+        # Returns the symmetric key for the hmac-secret extension also known as CredRandom.
+
+        if not self.hmac_secret:
+            return None
+
+        from apps.common import seed
+
+        node = seed.derive_slip21_node_without_passphrase(
+            [b"SLIP-0022", self.id[0:4], b"hmac-secret", self.id]
+        )
+
+        return node.key()
+
+
+class U2fCredential(Credential):
+    def __init__(self) -> None:
+        super().__init__()
+        self.node = None  # type: Optional[bip32.HDNode]
+
+    def __lt__(self, other: "Credential") -> bool:
+        # Sort U2F credentials after FIDO2 credentials.
+        if isinstance(other, Fido2Credential):
+            return False
+
+        # Sort U2F credentials lexicographically amongst each other.
+        return self.id > other.id
```

Comment:
There should be less-that sign instead of great-than sign.

Reply:
Sure.